R/images.R

```diff
@@ -0,0 +1,204 @@
+#' Capture a plot as a saved file
+#'
+#' Easily generates a .png file (or other graphics file) from a plotting
+#' expression.
+#'
+#' @param expr A plotting expression that generates a plot (or yields an object
+#'   that generates a plot when printed, like a ggplot2). We evaluate this
+#'   expression after activating the graphics device (\code{device}).
+#' @param filename The output filename. By default, a temp file with \code{.png}
+#'   extension will be used; you should provide a filename with a different
+#'   extension if you provide a non-PNG graphics device function.
+#' @param device A graphics device function; by default, this will be either
+#'   \code{\link[grDevices:png]{grDevices::png()}},
+#'   \code{\link[ragg:agg_png]{ragg::agg_png()}}, or
+#'   \code{\link[Cairo:CairoPNG]{Cairo::CairoPNG()}}, depending on
+#'   your system and configuration. See \code{\link{defaultPngDevice}}.
+#' @param width,height,res,... Additional arguments to the \code{device} function.
+#'
+#' @seealso \code{\link{plotTag}} saves plots as a self-contained \code{<img>}
+#'   tag.
+#'
+#' @examples
+#'
+#' # Default settings
+#' res <- capturePlot(plot(cars))
+#'
+#' if (interactive()) browseURL(res)
+#'
+#' # Use custom width/height
+#' pngpath <- tempfile(fileext = ".png")
+#' capturePlot(plot(pressure), pngpath, width = 800, height = 375)
+#'
+#' if (interactive()) browseURL(pngpath)
+#'
+#' # Use SVG
+#' svgpath <- capturePlot(
+#'   plot(pressure),
+#'   tempfile(fileext = ".svg"),
+#'   grDevices::svg,
+#'   width = 8, height = 3.75)
+#'
+#' if (interactive()) browseURL(svgpath)
+#'
+#' # Clean up
+#' unlink(res)
+#' unlink(pngpath)
+#' unlink(svgpath)
+#'
+#' @export
+capturePlot <- function(expr, filename = tempfile(fileext = ".png"),
+  device = defaultPngDevice(), width = 400, height = 400, res = 72,
+  ...) {
+  if (!is.function(device)) {
+    stop(call. = FALSE, "The `device` argument should be a function, e.g. `grDevices::png`")
+  }
+
+  expr <- rlang::enquo(expr)
+
+  tempFile <- missing(filename)
```

Comment:
This won't work if a `tempfile()` is supplied by the user. That said, I don't see a great way around this issue either. Maybe we could add a `cleanup` arg that could be used to explicitly request that we remove the `filename` if the expression fails?

Reply:
Hmmm... or maybe unconditionally remove it? I mean, by passing a path to `capturePlot`, clearly you don't care about what's currently at that path...

Reply:
Although equally, I don't think it's unreasonable that you called `capturePlot` with an explicit path, it failed, and a half-written file was left behind. The problem is that if `missing(filename)`, you don't even know what filename was used!

Reply:
Yeah... I think I like this as-is.

Reply:
You could add an argument `cleanup = missing(filename)`. That way, it would have the same default behavior as it currently does, but the user could change it if they wanted to.